### PR TITLE
Turn Restart Policy string value sets into enums

### LINKED ISSUE #493: Turn Restart Policy string value sets into enums
There are a few cases like the following: https://github.com/appvia/kev/blob/5a1c05c568427024fb3ef76a9fc0a5d6bdd1e341/pkg/kev/config/extensions.go#L187

Where the return value is a string and the possible values can only be obtained by reading the code and following all the different options. 

This could be transformed into Enums and supported by a few additional functions which ensure that conversion using different cases works (eg None, noNe, etc) and that there is a Valid() function that can quickly assert validity of a specific string value.

This ticket deals with Restart Policy.
1. Restart Policy

### PR DESCRIPTION
**Pull request by ezodude**
Resolves #493 

- Moved Restart Policy to use enums and clarified inference logic.
- Updated the render phase to use the svc K8s config value directly.
- Added tests to check the various Restart Policy scenarios.

### CODE REVIEW

File: pkg/kev/config/svcext.go [hunk 1]
```diff
@@ -299,26 +304,18 @@ func getServiceType(serviceType string) (string, error) {
...
-
-	if policy == "unless-stopped" {
-		log.WarnWithFields(log.Fields{
-			"restart-policy": policy,
-		}, "Restart policy 'unless-stopped' is not supported, converting it to 'always'")
-
-		policy = "always"
+func WorkloadRestartPolicyFromCompose(svc *composego.ServiceConfig) RestartPolicy {
+	switch {
+	case svc.Restart != "": // docker-compose v2
```
Comment (marcinc) [inline pkg/kev/config/svcext.go:309]: I don't think it's specific to v2, is it? Also, we should cascade here i.e. the value defined in `Deploy` block should override the value at the top level `Restart`.

File: pkg/kev/config/restart.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	case "", "always", "any":
+		return RestartPolicyAlways
+	case "no", "none", "never":
+		return RestartPolicyNever
+	case "on-failure", "onfailure":
+		return RestartPolicyOnFailure
+	case "unless-stopped":
+		return RestartPolicyAlways
+	default:
+		return ""
```
Comment (marcinc) [inline pkg/kev/config/restart.go:79]: this should perhaps return a default `DefaultRestartPolicy` rather than empty string.

File: pkg/kev/config/svcext_test.go [hunk 3]
```diff
@@ -85,88 +82,157 @@ var _ = Describe("Service Extension", func() {
...
-				})
-			})
-
-			Context("invalid/empty workload", func() {
-				JustBeforeEach(func() {
-					svc.Extensions = map[string]interface{}{
-						"x-k8s": map[string]interface{}{
-							"workload": map[string]interface{}{
+			Context("from svc to k8s ext", func() {
+				Context("with missing workload", func() {
```
Comment (marcinc) [inline pkg/kev/config/svcext_test.go:86]: you meant `with missing workload configuration`?

File: pkg/kev/config/svcext_test.go [hunk 4]
```diff
@@ -85,88 +82,157 @@ var _ = Describe("Service Extension", func() {
...
+					It("returns defaults", func() {
+						defaultWorkload := config.DefaultSvcK8sConfig().Workload
+						Expect(parsedK8sCfg.Workload).To(BeEquivalentTo(defaultWorkload))
+					})
 				})
 
-				When("workload is invalid", func() {
-					It("it is ignored and returns defaults", func() {
-						Expect(parsedK8sCfg).To(BeEquivalentTo(config.DefaultSvcK8sConfig()))
+				Context("invalid/empty workload", func() {
```
Comment (marcinc) [inline pkg/kev/config/svcext_test.go:101]: tiny nitpick, it'd be nice to keep the same wording as in previous context: `with invalid/empty workload configuration`. Should we tweak other context titles below in line with it it?

File: pkg/kev/converter/kubernetes/project_service.go [hunk 5]
```diff
@@ -322,22 +322,22 @@ func (p *ProjectService) serviceAccountName() string {
...
-	restartPolicy, err := getRestartPolicy(p.Name, policy)
-	if err != nil {
-		log.WarnWithFields(log.Fields{
-			"project-service": p.Name,
-			"restart-policy":  policy,
-		}, "Restart policy is not supported, defaulting to 'Always'")
-
-		return config.RestartPolicyAlways
-	}
+	// policy := config.RestartPolicyAlways
```
Comment (marcinc) [inline pkg/kev/converter/kubernetes/project_service.go:325]: comments..

File: pkg/kev/testdata/init-default/docker-compose-yml/docker-compose.yml [hunk 6]
```diff
@@ -1,4 +1,4 @@
-version: '3.9'
+version: '2.4'
```
Comment (marcinc) [inline pkg/kev/testdata/init-default/docker-compose-yml/docker-compose.yml:1]: Why this change?

**Review by marcinc — COMMENTED** [review]
A few comments. LGTMO. Also, liking the tweaks and cleanup to the test coverage 👍

File: pkg/kev/config/svcext.go (hunk 1 above)
Comment (ezodude) [reply pkg/kev/config/svcext.go:309]: Nope ... Its a v2 vs v3 thing. No cascade required. Hence the re-implementation.

v2: https://docs.docker.com/compose/compose-file/compose-file-v2/#restart
v3: https://docs.docker.com/compose/compose-file/compose-file-v3/#restart_policy

File: pkg/kev/testdata/init-default/docker-compose-yml/docker-compose.yml (hunk 6 above)
Comment (ezodude) [reply pkg/kev/testdata/init-default/docker-compose-yml/docker-compose.yml:1]: `restart: always` is a v2 docker-compose type config not v3.

> deploy.restart_policy  ... Configures if and how to restart containers when they exit. Replaces restart.
See here: https://docs.docker.com/compose/compose-file/compose-file-v3/#restart_policy

File: pkg/kev/config/svcext.go (hunk 1 above)
Comment (marcinc) [reply pkg/kev/config/svcext.go:309]: I asked as `Restart` is also possible in [V3 manifest](https://docs.docker.com/compose/compose-file/compose-file-v3/#restart), and thus we should still cascade I think.

File: pkg/kev/testdata/init-default/docker-compose-yml/docker-compose.yml (hunk 6 above)
Comment (marcinc) [reply pkg/kev/testdata/init-default/docker-compose-yml/docker-compose.yml:1]: see my comment below. `Restart` is still a valid v3 field.

File: pkg/kev/config/svcext.go (hunk 1 above)
Comment (ezodude) [reply pkg/kev/config/svcext.go:309]: That's odd the spec states otherwise.

Comment (ezodude) [reply pkg/kev/config/svcext.go:309]: Corrected.

File: pkg/kev/converter/kubernetes/project_service.go [hunk 7]
```diff
@@ -320,24 +321,18 @@ func (p *ProjectService) serviceAccountName() string {
...
-			"restart-policy":  policy,
-		}, "Restart policy is not supported, defaulting to 'Always'")
+// restartPolicy returns workload restart policy
+func (p *ProjectService) restartPolicy() (v1.RestartPolicy, error) {
+	return toV1RestartPolicy(p.SvcK8sConfig.Workload.RestartPolicy)
+}
 
-		return config.RestartPolicyAlways
+// toV1RestartPolicy maps to a case-sensitive v1 restart policy
+func toV1RestartPolicy(rp config.RestartPolicy) (v1.RestartPolicy, error) {
```
Comment (marcinc) [inline pkg/kev/converter/kubernetes/project_service.go:330]: should this func be moved to `restart.go`? I think this func could just take any previously inferred or config extracted RestartPolicy value and ensure that it returns a valid v1.RestartPolicy. In case the current config value wasn't correct perhaps we could just fall back to default instead of erroring? Though I get why you'd want to error in some cases... What do you think?

**Review by marcinc — COMMENTED** [review]
one last comment.

File: pkg/kev/converter/kubernetes/project_service.go (hunk 7 above)
Comment (ezodude) [reply pkg/kev/converter/kubernetes/project_service.go:330]: Funny you should ask this question - because that's where the func started 😄 ...

But I moved it here because its a converter issue. Different converters may need to deal with this value in different ways.

Regarding the error, the likelihood is small but it's a useful guard. Also, useful for a user to rectify any bad config.

**Review by marcinc — APPROVED** [review]
LG
